tokio/src/sync/broadcast.rs

```diff
@@ -675,42 +640,28 @@ impl<T> Drop for Sender<T> {
 
 impl<T> Receiver<T> {
     /// Locks the next value if there is one.
-    ///
-    /// The caller is responsible for unlocking
-    fn recv_ref(&mut self, spin: bool) -> Result<RecvGuard<'_, T>, TryRecvError> {
+    fn recv_ref(&mut self) -> Result<RecvGuard<'_, T>, TryRecvError> {
         let idx = (self.next & self.shared.mask as u64) as usize;
 
         // The slot holding the next value to read
-        let slot = &self.shared.buffer[idx];
-
-        // Lock the slot
-        if !slot.try_rx_lock() {
-            if spin {
-                while !slot.try_rx_lock() {
-                    spin_loop_hint();
-                }
-            } else {
-                return Err(TryRecvError::Empty);
-            }
-        }
-
-        let guard = RecvGuard {
-            slot,
-            tail: &self.shared.tail,
-            condvar: &self.shared.condvar,
-        };
-
-        if guard.pos() != self.next {
-            let pos = guard.pos();
+        let mut slot = self.shared.buffer[idx].read().unwrap();
 
+        if slot.pos != self.next {
             // The receiver has read all current values in the channel
-            if pos.wrapping_add(self.shared.buffer.len() as u64) == self.next {
-                guard.drop_no_rem_dec();
+            if slot.pos.wrapping_add(self.shared.buffer.len() as u64) == self.next {
+                // guard.drop_no_rem_dec();
                 return Err(TryRecvError::Empty);
             }
 
+            // In order to avoid deadlocks, release the `slot` lock before
+            // attempting to acquire the `tail` lock.
+            drop(slot);
```

Comment:
Not sure I follow why the slot needs to be dropped to be relocked again after locking the tail.. could this comment clarify that a little more?

Reply:
I expanded on the comment.

> Release the `slot` lock before attempting to acquire the `tail`
> lock. This is required because `send2` acquires the tail lock
> first followed by the slot lock. Acquiring the locks in reverse
> order here would result in a potential deadlock: `recv_ref`
> acquires the `slot` lock and attempts to acquire the `tail` lock
> while `send2` acquired the `tail` lock and attempts to acquire
> the slot lock.